README.md
# MongoDB YCSB Charts

## Introduction
This program creates a bubble chart. The y axis is the throughput as measured in ops / sec. The x axis shows number of threads the trials are run with. 
The size of the bubble is the absolute difference in proportions among read/update/scan/insert parameters (this is more of a visual aid to estimate the range of the proportions, by far not a scientific proxy). 
The color of the bubble denotes the combination of its storage engine, mongod version, and workload ratio.

## How to run
Make sure everything required of YCSB is installed (e.g. Maven and Java JDK). If running for the first time after cloning this repo, run ```mvn clean package```. Next, ensure that Python 2.7, pandas, appJar, matplotlib, and numpy are installed. Then run ```python create_chart.py``` to start.

## What works
Through a GUI, the script takes in several settings for running YCSB, how many trials you'd like to compare, and the settings for each trial. The script can also compare different mongod versions (binary files besides 3.4.7 and 3.5.10 need to be added in manually) and storage engines. Workload files are created and saved, then each workload file is loaded and executed by YCSB through the command line for all specified mongod / storage engine combinations. Each .out file resulting from these executions is parsed for the throughput metric, and the number is saved in an accumulating dictionary the program keeps for the duration of its lifetime. At the end, all throughput data points are fed into a bubble chart, and the resulting bubble chart appears.

This program has been tested on Mac OSX 10.12.

An example run is as follows:  

![Collecting Options](https://raw.githubusercontent.com/kimberlyhou/mongodb-ycsb-charts/master/images/options_screenshot.png)
   
After executing all workload files, the program prints out a dictionary of the data with settings and the exact numeric throughputs organized by thread group:
![Logged Output](https://raw.githubusercontent.com/kimberlyhou/mongodb-ycsb-charts/master/images/stdout_screenshot.png)
   
The resulting bubble chart:
![Bubble Chart](https://raw.githubusercontent.com/kimberlyhou/mongodb-ycsb-charts/master/images/bubble_chart.png)

The record count and operation count are two of the biggest indicators for how long a run would take. With 1 million each, this trial took 36 minutes to run. If they were in the low thousands, it would take less than 1 minute. See the next section for more details.

## What needs work
The following known issues exist: 
First, the more trials, workload options, and mongod versions/storage engines you want to compare, the more patient you'd have to be - exponentially. While comparing 1-2 trials and 1-2 mongod versions may take a couple minutes to create the bubble chart, the entire operation could take upwards of 1 hour 20 minutes to eventually create the bubble chart if all the options are maxed out (e.g. 4 workload ratios, 8 thread groups, and 2 mongod versions each are compared). The slowdown occurs when MongoDB is writing to the log after executing each YCSB workload file.
Keep in mind YCSB has to load and run all the different workload ratios for all specified thread # groups for every unique mongod/storage engine combination. The solution to this could be to create mongod .conf files ahead of time and add them to the repo, allowing the Python script to run mongod based on these configuration files instead of just via the command line. In particular, the configuration files could have settings such as systemLog: quiet: true to limit the amount that MongoDB logs. These settings could add robustness from forking failure and potentially increase speed of the overall operation.  

Second, if the max execution time is too low or YCSB encounters a one-off problem with loading / running a workload file, this program isn't robust enough to handle such situations, resulting in a hang. Exiting the program after a set timeout limit is reached or restarting a failed command based on the error would help. 

Third, depending on how the bubble chart comes out, the legend may need to be repositioned in order to see all the bubbles. If the entire program runs and the bubble chart comes out partially blocked by the legend (even after enlarging the window), a workaround for the current version of the script is to take the final dictionary of data printed to stdout after all the executions and then feed it into the create_bubble_chart function, running only that function after tweaking the legend position parameters. Probably the best solution is to shrink the entire graph and move the legend outside the plot without having the window cut it off.

## Further information
This repo is based on the 10gen/MongoDB maintained partial YCSB repo with a few additional files, mainly create_chart.py and collect_settings.py. The files here should be mergable with the original project if necessary.  These files can be built and run standalone, but only MongoDB client and core YCSB libraries are provided.

For more details, see https://github.com/10gen-labs/YCSB/wiki and https://github.com/mongodb-labs/YCSB/blob/master/ycsb-mongodb/mongodb/README.md.


CHANGELOG
-95 Bump MongoDB version to 2.9.0 (allanbank)
- gh-67 Use checkstyle (m1ch1)
- gh-76 Implemented OrientDB client (lvca)
- gh-88 YCSB client for Amazon DynamoDB (jananin)
- gh-89 Patch for YCSB Cassandra Client version 1.0.6 (jananin)
- gh-93 New ElasticSearch Database Implementation (saden1)
- gh-97 Bug fixes in dynamodb plugin (jananin)

0.1.4 - 2/22/12 

- Fixes for Cassandra 0.7 client (tjake)
- New generator FileGenerator (nono)
- Fixes for MongoDB (nono)
- Fixes for Voldemort (nono)
- JDBC client (sudiptodas)
- HotspotIntegerGenerator (sudiptodas)	
- Optimizing cassandra7 (joaquincasares)
- Mysql key fix (joaquincasares)
- Added a db plugin for Infinispan (maniksurtani)
- gh-31 Support to stop benchmark after a maximum specified elapsed time. (sudiptodas)
- gh-35 Cassandra0.8 support (joaquincasares)
- gh-30 IllegalArgumentException with MongoDB (m1ch1)
- gh-27 MongoDbClient was not working with non localhost URLs (arunxarun)
- gh-29 Add simple sharding capabilities for JDBC driver (kibab)
- gh-40 Merge Redis database interface layer (lehmannro)
- gh-42 Response latencies are measured in microseconds (mikewied)
- gh-43 Variable length fields (sears)
- gh-44 Constant occupancy workload (sears)
- gh-45 Modify DB API for efficient large object support (sears)
- gh-46 Fixed typo in RedisClient (Zlender)
- gh-49 Build fix (sears)
- gh-50 Switch unordered key generation from FNV32 to FNV64 (sears)
- gh-51 Improved Random Number Generation Performance and add Exponential distribution support (sears)
- gh-52 Mongo db fix (sears)
- gh-54 Add mapkeeper driver (m1ch1)
- gh-57 voldemort - enable nio connector (akkumar)
- gh-58 benchmarking with hbase 0.90.5 (akkumar)
- gh-55 VMware vFabric GemFire (sbawaska)
- gh-59 Cassandra 1.0.6 (akkumar)
- gh-62 Oracle NoSQL Database Client (y-namiki)
- gh-64 Mavenisation of YCSB with a (tar ball) distribution packager (hariprasad-k)
- gh-65 Patch for improving the MongoDB Client (singhsiddharth)

0.1.3 - 10/26/10

= Voldemort binding (rsumbaly)
= HBase client improvements (ryanobjc)
= Fixes to Cassandra 0.7 binding (johanoskarsson, nickmbailey)
= Added an interface for exporting the measurements and a JSON implementation. It can write to both stdout and to a file (johanoskarsson)
-Other minor fixes (brianfrankcooper)

0.1.2 - 5/12/10

- MongoDB binding (ypai)
- Cassandra 0.7 binding (johanoskarsson)
- Simple command line interface (brianfrankcooper)
- Faster string generation (tlipcon)
- Avoid Bytes conversion in HBaseClient (tlipcon)

0.1.1 - 4/25/10

- Compiles under 1.5
- Fixes doc and HBaseClient bugs

0.1.0 - 4/23/10 

- Initial open source release


create_chart.py
'''This script contains the definitions for executing workload files using ycsb
and then parsing the throughput datapoints to visualize on a bubble chart'''

from subprocess import check_output, CalledProcessError, STDOUT

import matplotlib
matplotlib.use('TkAgg')
from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
import sys

from collect_settings import open_gui

def exec_cmd(cmd):
    '''Wrapper function'''
    check_output(cmd, stderr=STDOUT, shell=True)

def load_and_run(dictionary):
    '''Load and run workload files'''
    # Start mongod
    mongod_se = []
    throughputs = {}
    # To prevent starting a mongod on the same port before the previous mongod fully stopped,
    # a temporarily solution is to start each mongod on different ports to prevent an 
    # errno 48 upon failed --fork.
    port = 27018
    for mongod in dictionary['mongod_versions']:
        print 'mongod is {}'.format(mongod)
        throughputs[mongod] = {}
        for storage_engine in dictionary['storage_engines']:
            print 'storage_engine is {}'.format(storage_engine)
            mongod_se.append('{} {}'.format(mongod, storage_engine))
            throughputs[mongod][storage_engine] = {}
            exec_cmd('pgrep mongod | xargs kill')
            print 'just killed existing mongod processes'
            # 'killall -w mongod' on Linux
            try:
                # Future work: Have the user input a config file to setup mongod
                exec_cmd('./bin/mongod-{} --fork --logpath /dev/null --dbpath data/{} --port {} --storageEngine {}'.format(\
                         mongod, storage_engine, port, storage_engine))
            except CalledProcessError as err:
                print err
                print err.output
                sys.exit(1)

            # initialize empty dictionaries
            for thread in dictionary['threads']:
                throughputs[mongod][storage_engine][thread] = {}

            for workload_file_group in range(len(dictionary['workload_files'])):
                threadgroup = dictionary['threads'][workload_file_group]
                throughputs[mongod][storage_engine][threadgroup] = []
                for workload_file in dictionary['workload_files'][workload_file_group]:
                    print 'loading {}...'.format(workload_file)
                    exec_cmd('./bin/ycsb load mongodb -s -P {} -p mongodb.url=localhost:{}'.format(workload_file, port))
                    print 'running {}...'.format(workload_file)
                    exec_cmd('./bin/ycsb run mongodb -s -P {} -p mongodb.url=localhost:{}'.format(workload_file, port))
                    result = parse_throughput(workload_file)
                    throughputs[mongod][storage_engine][threadgroup].append(result)

            print 'done, going to kill this mongod'
            exec_cmd('pgrep mongod | xargs kill')
            port += 1

    dictionary['mongod_se'] = mongod_se
    dictionary['throughputs'] = throughputs

def parse_throughput(workload_file):
    '''Return the throughput number from an .out file'''
    working_file = open(workload_file + '.out')
    line = working_file.readline()
    while 'Throughput' not in line:
        line = working_file.readline()
    line = line.split()
    working_file.close()
    return float(line[-1])

def ratios_to_label(dictionary, index, label):
    '''Add and format the workload ratios to a given label'''
    static_workloads = ['read', 'update', 'scan', 'insert']
    for rusi in static_workloads:
        if dictionary['workload_ratios'][index][rusi] > 0:
            # At this scale, concatenation in Python 2.4+ can be faster than append/join.
            label += '{}: {} '.format(rusi.title(), \
                     str(int(dictionary['workload_ratios'][index][rusi]*100)))
    return label

def choose_color(index):
    '''Cycle through an array of colors'''
    colors = ['#7e1e9c', '#15b01a', '#0343df', '#e50000', '#029386', '#c20078', '#75bbfd',
              '#06470c', '#9a0eea', '#840000', '#b9a281', '#040273', '#fc5a50', '#5170d7',
              '#1fa774']
    if index < len(colors):
        return colors[index]
    else:
        return colors[index / len(colors)]

def create_bubble_chart(dictionary):
    '''Feed the data points into the bubble chart'''
    x_axis = []
    for thread in dictionary['threads']:
        x_axis.append(int(thread))
    dataframe_data = {}
    labels = []
    y_s_axes = []
    for mongod in dictionary['mongod_versions']:
        for storage_engine in dictionary['storage_engines']:
            for workload_index in range(len(dictionary['workload_ratios'])):
                throughputs = []
                label = ''
                label = ratios_to_label(dictionary, workload_index, label)
                label += '- {} {}'.format(mongod, storage_engine)
                labels.append(label)
                for thread in dictionary['threads']:
                    throughputs.append(int(dictionary['throughputs'][mongod][storage_engine]\
                                       [thread][workload_index]))
                bubble_size_arr = [dictionary['workload_ranges'][workload_index]] * \
                                  len(dictionary['threads'])
                y_s_axes.append((np.array(throughputs), np.array(bubble_size_arr)))

    # Collect all the y and s arrays into one dictionary for the dataframe
    for tuple_index in range(len(y_s_axes)):
        dataframe_data['{}.y'.format(tuple_index)] = y_s_axes[tuple_index][0]
        dataframe_data['{}.s'.format(tuple_index)] = y_s_axes[tuple_index][1]
    dataframe_data['x'] = np.array(x_axis)
    dataframe_columns = list(dataframe_data.keys())
    dataframe = pd.DataFrame(dataframe_data, columns=dataframe_columns)

    # Create the plot group by group
    for tuple_index in range(len(y_s_axes)):
        if tuple_index == 0:
            axes = dataframe.plot(kind='scatter', x='x', y='{}.y'.format(tuple_index), \
                 s=(dataframe['{}.s'.format(tuple_index)]+2)*15, color=choose_color(tuple_index), \
                 label=labels[tuple_index], alpha=0.5)
        else:
            dataframe.plot(kind='scatter', x='x', y='{}.y'.format(tuple_index), \
                 s=(dataframe['{}.s'.format(tuple_index)]+2)*15, color=choose_color(tuple_index), \
                 label=labels[tuple_index], ax=axes, alpha=0.5)

    axes.set_xlabel("# threads")
    axes.set_ylabel("Throughput (ops / sec)")
    plt.legend(loc='upper center', bbox_to_anchor=(0.5, 1.05), fancybox=True, shadow=True)
    plt.show()

if __name__ == "__main__":
    data = open_gui()
    print 'executing workload files...'
    load_and_run(data)
    print data
    print 'finished loading / running workload files'
    create_bubble_chart(data)


poc_create_chart.py
import numpy as np
import matplotlib.pyplot as plot
from subprocess import check_output, CalledProcessError, STDOUT

# load and run the two files
# for each file, parse the output for [THROUGHPUT] metric
# save that number

def exec_cmd(cmd):
	check_output(cmd, stderr=STDOUT, shell=True)

def load_and_run():
	exec_cmd('pwd')
#	try:
	exec_cmd('./bin/ycsb load mongodb -s -P workloads/workloadb > outputLoadb.txt')
	exec_cmd('./bin/ycsb run mongodb -s -P workloads/workloadb > outputRunb.txt')
	exec_cmd('./bin/ycsb load mongodb -s -P workloads/workloadc > outputLoadc.txt')
	exec_cmd('./bin/ycsb run mongodb -s -P workloads/workloadc > outputRunc.txt')
#	except CalledProcessError as e:
#		print(e.output)

def parse_output_files():
	with open('outputRunb.txt') as b:
		lines = b.readlines()
		for line in lines:
			if 'RunTime' in line:
				line = line.split()
				v1_runtime = line[-1]
			if 'Throughput' in line:
				line = line.split()
				v1_throughput = line[-1]
	with open('outputRunc.txt') as b:
		lines = b.readlines()
		for line in lines:
			if 'RunTime' in line:
				line = line.split()
				v2_runtime = line[-1]
			if 'Throughput' in line:
				line = line.split()
				v2_throughput = line[-1]
	v1 = [float(v1_runtime), float(v1_throughput)]
	v2 = [float(v2_runtime), float(v2_throughput)]
	return [v1, v2]

def create_chart(results):
	n_groups = len(results)
	v1_data = (results[0][0], results[0][1])
	v2_data = (results[1][0], results[1][1])
	print(type(v1_data))

	fig, ax = plot.subplots()
	index = np.arange(n_groups)
	bar_width = 0.15
	opacity = 0.8

	rects1 = plot.bar(index, v1_data, bar_width,
                 alpha=opacity,
                 color='#d6b2ff',
                 label='Workload b',
                 align='edge',
                 linewidth=0.8)

	rects2 = plot.bar(index + bar_width, v2_data, bar_width,
                 alpha=opacity,
                 color='#a6ddff',
                 label='Workload c',
                 align='edge')

	plot.xlabel('Workload')
	plot.ylabel('Metrics')
	plot.title('Metrics by Workload')
	plot.xticks(index + bar_width, ('Runtime (m/s)', 'Throughput (ops/sec)'))
	plot.legend()
	plot.tight_layout()
	plot.show()

if __name__ == '__main__':
	load_and_run()
	create_chart(parse_output_files())


collect_settings.py
'''This script contains the definitions for showing a GUI to the user
in order to collect options the user would like to select for a given
comparison'''

from appJar import gui

# Future: Save settings to a txt working_file and enable reloading of a setting from that working_file

def open_gui():
    '''Show GUI to user'''
    result_object = {}

    def start_gui():
        '''Entrypoint function'''
        app.addLabel('label_dimensions', \
                     'Welcome! How many workload ratios would you like to compare?')
        app.addOptionBox('dimension_comparisons', range(1, 5))
        app.addButton('Next', specify_dim_comparisons)
        app.go()

    def specify_dim_comparisons(btn):
        '''Second step'''
        app.removeButton('Next')
        app.disableOptionBox('dimension_comparisons')
        num_dimensions = int(app.getOptionBox('dimension_comparisons'))

        # Ensure the labeled mongod binary is in the bin/ directory of this repo.
        for i in range(num_dimensions):
            app.addLabelNumericEntry('Read Proportion #{}'.format(i))
            app.addLabelNumericEntry('Update Proportion #{}'.format(i))
            app.addLabelNumericEntry('Scan Proportion #{}'.format(i))
            app.addLabelNumericEntry('Insert Proportion #{}'.format(i))
            app.setEntryDefault('Read Proportion #{}'.format(i), 0.25)
            app.setEntryDefault('Update Proportion #{}'.format(i), 0.25)
            app.setEntryDefault('Scan Proportion #{}'.format(i), 0.25)
            app.setEntryDefault('Insert Proportion #{}'.format(i), 0.25)

        app.addLabel('label_se',
                     'Choose at least one storage engine to compare:')
        storage_engines = {'wiredTiger' : False, 'mmapv1' : False}
        app.addProperties('Storage engines', storage_engines)
        app.addLabel('label_mongod',
                     'Choose at least one mongod version to compare:')
        mongod_versions = {'3.4.7' : False, '3.5.10' : False}
        app.addProperties('Mongod versions', mongod_versions)

        app.addButton('Next', choose_num_trials)

    def choose_num_trials(btn):
        '''Choose # groups and # threads per group'''
        app.removeButton('Next')
        app.addLabel('label_num_trials', \
             'How many trials would you like to compare? \
             Each trial has a different # of threads.')
        app.addOptionBox('num_trials', ['- # trials -', 1, 2, 3, 4, 5, 6, 7, 8])
        app.addButton('Next', specify_thread_comparisons)

    def specify_thread_comparisons(btn):
        '''Show x boxes to input threads per group'''
        app.removeButton('Next')
        app.disableOptionBox('num_trials')
        num_trials = int(app.getOptionBox('num_trials'))
        for i in range(num_trials):
            app.addLabel('label_num_threads_group' + str(i), '# threads for group ' + str(i))
            app.addOptionBox('num_threads_group' + str(i), range(1, 25))
        show_remaining_options()

    def show_remaining_options():
        '''Remaining options are fixed'''
        app.addLabelNumericEntry('Field Count')
        app.setEntryDefault('Field Count', 50)
        app.addLabelNumericEntry('Field Length')
        app.setEntryDefault('Field Length', 10)
        app.addLabelNumericEntry('Record Count')
        app.setEntryDefault('Record Count', 5000000)
        app.addLabelNumericEntry('Operation Count')
        app.setEntryDefault('Operation Count', 10000000)
        app.addLabelNumericEntry('Max Execution Time')
        app.setEntryDefault('Max Execution Time', 5000)
        app.addButton('Run', validate_before_write)

    def check_proportions(num_dimensions):
        '''Validate and add ratios to global variable result_object'''
        workload_ratios = []
        for i in range(num_dimensions):
            read = app.getEntry('Read Proportion #{}'.format(i))
            update = app.getEntry('Update Proportion #{}'.format(i))
            scan = app.getEntry('Scan Proportion #{}'.format(i))
            insert = app.getEntry('Insert Proportion #{}'.format(i))
            if read + update + scan + insert != 1:
                app.errorBox('errorWorkload', 'Error: Proportions have to add up to 1')
                return False
            workload_ratios.append(dict(read=read, update=update, scan=scan, insert=insert))
        result_object['workload_ratios'] = workload_ratios
        return True

    def validate_threads_input(num_trials):
        '''Validate before returning threads'''
        threads = []
        # Collect number of threads selected for each run with validation check
        for i in range(num_trials):
            threads.append(app.getOptionBox('num_threads_group' + str(i)))
        if len(threads) != len(set(threads)):
            revise = app.yesNoBox('warningDuplicateTrials',
                                  'Warning: At least one of your trials is a duplicate.'\
                                  'Would you like to revise?')
            if revise:
                return (False, [])
        return (True, threads)

    def check_properties(prop_name):
        '''Validate before returning properties'''
        dictionary = app.getProperties(prop_name)
        lst = []
        for key, value in dictionary.iteritems():
            if value:
                lst.append(key)
        if len(lst) == 0:
            app.errorBox('error_{}'.format(prop_name),\
                         'Error: Please select at least one property for {}.'.format(prop_name))
            return (False, [])
        return (True, lst)

    def validate_before_write(btn):
        '''When the run button is pressed, validate before writing'''
        app.disableButton('Run')
        num_trials = int(app.getOptionBox('num_trials'))
        num_dimensions = int(app.getOptionBox('dimension_comparisons'))
        check_threads = validate_threads_input(num_trials)
        check_se = check_properties('Storage engines')
        check_mongod = check_properties('Mongod versions')
        valid = \
            check_threads[0] and \
            check_se[0] and \
            check_mongod[0] and \
            check_proportions(num_dimensions)
        if valid:
            write_workload_files(num_trials, num_dimensions, check_threads[1], \
                                 check_se[1], check_mongod[1])
        else:
            app.enableButton('Run')

    def write_workload_files(num_trials, num_dimensions, threads, storage_engines, mongods):
        '''Create workload files for all specified settings and set result_object'''
        field_count = int(app.getEntry('Field Count'))
        field_length = int(app.getEntry('Field Length'))
        record_count = int(app.getEntry('Record Count'))
        operation_count = int(app.getEntry('Operation Count'))
        max_exec_time = int(app.getEntry('Max Execution Time'))

        workload_files = []
        workload_labels = []
        workload_ranges = []

        for thread in range(num_trials):
            files_threadgroup = []
            for dim in range(num_dimensions):
                read = result_object['workload_ratios'][dim]['read']
                update = result_object['workload_ratios'][dim]['update']
                scan = result_object['workload_ratios'][dim]['scan']
                insert = result_object['workload_ratios'][dim]['insert']
                workload_labels.append('RUSI: {}-{}-{}-{}'.format(read, update, scan, insert))
                absolute_range = abs(abs(abs(int(read * 100) - int(update * 100)) - int(scan * 100)) - int(insert * 100))
                workload_ranges.append(absolute_range)

                filename = 'fc{}fl{}rc{}-r{}u{}s{}i{}-t{}'.format(field_count,
                    field_length, record_count, int(read * 100), int(update * 100),
                    int(scan * 100), int(insert * 100), threads[thread])

                working_file = open(filename, 'w')
                working_file.write('fieldcount={}\n'.format(field_count))
                working_file.write('fieldlength={}\n'.format(field_length))
                working_file.write('recordcount={}\n'.format(record_count))
                working_file.write('operationcount={}\n'.format(operation_count))
                working_file.write('maxexecutiontime={}\n'.format(max_exec_time))
                working_file.write('threadcount={}\n'.format(threads[thread]))
                working_file.write('workload=com.yahoo.ycsb.workloads.CoreWorkload\n')
                working_file.write('exportfile={}\n'.format(filename + '.out'))
                working_file.write('readallfields=true\n\n')
                working_file.write('readproportion={}\n'.format(read))
                working_file.write('updateproportion={}\n'.format(update))
                working_file.write('scanproportion={}\n'.format(scan))
                working_file.write('insertproportion={}\n\n'.format(insert))
                working_file.write('requestdistribution=zipfian\n')
                working_file.close()
                files_threadgroup.append(filename)
                print '{} has been saved.'.format(filename)
            workload_files.append(files_threadgroup)

        result_object['storage_engines'] = storage_engines
        result_object['mongod_versions'] = mongods
        result_object['workload_files'] = workload_files
        result_object['workload_labels'] = workload_labels
        result_object['workload_ranges'] = workload_ranges
        result_object['threads'] = threads
        app.stop()

    app = gui()
    start_gui()
    return result_object
